frame/transaction-payment/rpc/runtime-api/src/lib.rs

```diff
@@ -31,4 +31,12 @@ sp_api::decl_runtime_apis! {
 		fn query_info(uxt: Block::Extrinsic, len: u32) -> RuntimeDispatchInfo<Balance>;
 		fn query_fee_details(uxt: Block::Extrinsic, len: u32) -> FeeDetails<Balance>;
 	}
+
+	pub trait TransactionPaymentCallApi<Balance, Call> where
+		Balance: Codec + MaybeDisplay,
+		Call: Codec,
```

Comment:
with separate trait for Call we don't need to define and pass the Call generic argument for all clients using rpc server buld with TransactionPaymentApi

Reply:
I am not sure if I agree with this. If we merge the two traits. I think having to implement a new trait is worse than needing to add one generic type that is used in two of the functions.

Reply:
The problem here being that a runtime api should always use a "stable" api. Meaning, you should only use types that are known to the node and to the runtime. For example the extrinsic is opaque to the client, but it has a fixed encoding that the client can decode. Here we should pass `Call` as a `Vec<u8>` and then decode it in the runtime manually.

Reply:
@bkchr let me share my understanding, and please correct me if I am wrong. 
There are currently two general approaches to call runtime api. 

1. Custom RPC server. 
This requires node and runtime to be in sync, specially for the types like `Call` (not a primitive).
With the approach we will pass extrinsic or call as encoded string to the rpc server, decode it manually, and forward the type to runtime api. [example](https://github.com/paritytech/substrate/blob/fdfdc73f9e64dc47934b72eb9af3e1989e4ba699/frame/transaction-payment/rpc/src/lib.rs#L104).

2. state_call (or api.rpc.state.call [endpoint](https://polkadot.js.org/docs/substrate/rpc/#state))
Does not require a node to have the same version of runtime with wasm binary. 
Runtime provides a metadata, which is used to decode/encode types.
Happens somewhere [here](https://github.com/paritytech/substrate/blob/20fb5d05a3df8a9335bad3752abfed827803cd1f/client/executor/wasmi/src/lib.rs#L221).

Since we introducing only runtime_api methods to be used via `state_call` we do not need to do decoding manually.
And Extrinsic above, `Block::Extrinsic` from `TransactionPaymentApi::query_info` is not opaque.

Reply:
@kianenigma @bkchr 
Here how it will look like if we use the same `TransactionPaymentApi` trait for the Call methods, it forces us to declare new generic where the trait is used. 
https://github.com/paritytech/substrate/commit/2dc9e4b133910c1a9633400be1c0d9a69424981a

I also think it's better for the maintenance. But wanted to avoid updating the node's codebase.

Reply:
I test different cases today locally

Reply:
Test results, 

First tried the solution where `TransactionPaymentApi` trait extended with methods and `Call` generic. 
This requires the type `Call` be in sync with node and runtime.
Here `Call` must be declared for node rpc server - [link](https://github.com/paritytech/polkadot/blob/7c8b07e246dfda458d68c90b44889297c7a92e21/rpc/src/lib.rs#L110)
I have not gone further with this solution. It does not fit conceptually and we are not adding custom RPC calls.

Second, I have tested the current PR, with the additional trait `TransactionPaymentCallApi`.
Build and run the relay node and collator with parachain from master. 
Build and uploaded a new runtime for parachain with these new api calls, and included new pallet (sudo) to change `Call` type. 
Invoked new `TransactionPaymentCallApi_query_call_info` api via rpc `state_call` with the call arg containing a call from new pallet.
And it works correct. 

I would propose to keep the current version with an additional `TransactionPaymentCallApi` trait.

Reply:
> @bkchr let me share my understanding, and please correct me if I am wrong. There are currently two general approaches to call runtime api.
> 
>     1. Custom RPC server.
>        This requires node and runtime to be in sync, specially for the types like `Call` (not a primitive).
>        With the approach we will pass extrinsic or call as encoded string to the rpc server, decode it manually, and forward the type to runtime api. [example](https://github.com/paritytech/substrate/blob/fdfdc73f9e64dc47934b72eb9af3e1989e4ba699/frame/transaction-payment/rpc/src/lib.rs#L104).
> 
>     2. state_call (or api.rpc.state.call [endpoint](https://polkadot.js.org/docs/substrate/rpc/#state))
>        Does not require a node to have the same version of runtime with wasm binary.
>        Runtime provides a metadata, which is used to decode/encode types.
>        Happens somewhere [here](https://github.com/paritytech/substrate/blob/20fb5d05a3df8a9335bad3752abfed827803cd1f/client/executor/wasmi/src/lib.rs#L221).
> 
> 
> Since we introducing only runtime_api methods to be used via `state_call` we do not need to do decoding manually. And Extrinsic above, `Block::Extrinsic` from `TransactionPaymentApi::query_info` is not opaque.

Even if we only use the `state_call` and we don't need this parameter in the node, it still breaks "use fixed types" for runtime apis. Please use some `Vec<u8>` and do the decoding "manually". This isn't really that hard and you can also just do it inside of the function in `transaction-payment`.

Reply:
@bkchr sure! I change it to Vec. 
Just wanted to understand it better, since initially I had a Vec instead of Call, and was asked to change it to Call.